tequilapi/http_api_server.go

```diff
@@ -0,0 +1,74 @@
+package tequilapi
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+type ApiServer interface {
+	Port() int
+	Wait()
+	Stop()
+}
+
+type apiServer struct {
+	listener  net.Listener
+	stopped   *sync.WaitGroup
+	boundPort int
+}
+
+func NewServer(address string, port int, handler http.Handler) (ApiServer, error) {
+	var err error
+	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", address, port))
+	if err != nil {
+		return nil, err
+	}
+
+	boundPort, err := extractBoundPort(listener)
+	if err != nil {
+		listener.Close()
+		return nil, err
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	server := apiServer{listener, &wg, boundPort}
+	go serve(listener, handler, &wg)
+	return &server, nil
+}
+
+func (server *apiServer) Stop() {
+	server.listener.Close()
+}
+
+func (server *apiServer) Wait() {
+	server.stopped.Wait()
```

Comment:
I don't quite understand how to use this `Wait` logic. what are we waiting for and why?

Reply:
A good start is here: https://www.golang-book.com/books/intro/10